strawberryfields/engine.py

```diff
@@ -86,15 +87,40 @@ class Result:
 
     Represents the results of the execution of a quantum program.
     Returned by :meth:`.BaseEngine.run`.
+
+    Args:
+        samples (list[list[int, float, None]]): nested list of
+            measurement samples, with shape ``(modes, shots)``
```

Comment:
How about storing the samples already in `BaseEngine` as a Dict[int, List[float,int]], and passing a copy of that to Result? I thought we agreed that storing Nones isn't the way to go, right?

Reply:
If the measurement samples obtained from the backend API call are stored in RegRefs first, transferring them into the samples dictionary only requires
`BaseEngine.samples = {r.ind: r.val for r in prog.reg_refs.values() if r.val is not None}`

I don't know how RegRefs would store batched data though.

Reply:
Yes, that's right about `None`s. I think @josh146 just forgot that changing that behaviour was in-scope for this issue as well

Reply:
> If the measurement samples obtained from the backend API call are stored in RegRefs first, transferring them into the samples dictionary only requires
> `BaseEngine.samples = {r.ind: r.val for r in prog.reg_refs.values() if r.val is not None}`
> 
> I don't know how RegRefs would store batched data though.

Yep, seems like this is the easiest way to do it. Presuming everything is returned from backend in good order, should be simple to convert to a dict and wrap in `Results`

Reply:
I believe batched data was just stored as an array in each RegRef